components/web_discovery/browser/ecdh_aes.cc

```diff
@@ -0,0 +1,139 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/web_discovery/browser/ecdh_aes.h"
+
+#include <array>
+
+#include "base/base64.h"
+#include "base/logging.h"
+#include "base/ranges/algorithm.h"
+#include "base/strings/string_number_conversions.h"
+#include "crypto/random.h"
+#include "crypto/sha2.h"
+#include "third_party/boringssl/src/include/openssl/aead.h"
+#include "third_party/boringssl/src/include/openssl/ec.h"
+#include "third_party/boringssl/src/include/openssl/ec_key.h"
+#include "third_party/boringssl/src/include/openssl/ecdh.h"
+#include "third_party/boringssl/src/include/openssl/nid.h"
+
+namespace web_discovery {
+
+namespace {
+
+constexpr size_t kAesKeySize = 16;
+constexpr size_t kAesTagLength = 16;
+constexpr size_t kIvSize = 12;
+constexpr size_t kKeyMaterialSize = 32;
+// P-256 field size * 2 + type byte
+constexpr size_t kComponentOctSize = 32 * 2 + 1;
+// type byte + public component + initialization vector
+constexpr size_t kEncodedPubKeyAndIv = 1 + kComponentOctSize + kIvSize;
+constexpr uint8_t kP256TypeByte = 0xea;
+
+bssl::UniquePtr<EC_KEY> CreateECKey() {
+  return bssl::UniquePtr<EC_KEY>(
+      EC_KEY_new_by_curve_name(NID_X9_62_prime256v1));
+}
+
+}  // namespace
+
+AESEncryptResult::AESEncryptResult(std::vector<uint8_t> data,
+                                   std::string encoded_public_component_and_iv)
+    : data(data),
+      encoded_public_component_and_iv(encoded_public_component_and_iv) {}
+
+AESEncryptResult::~AESEncryptResult() = default;
+AESEncryptResult::AESEncryptResult(const AESEncryptResult&) = default;
+
+std::optional<AESEncryptResult> DeriveAESKeyAndEncrypt(
+    const std::string& server_pub_key_b64,
+    const std::vector<uint8_t>& data) {
+  auto server_pub_key_data = base::Base64Decode(server_pub_key_b64);
```

Comment:
it seems strange to me that we're doing this low level encryption stuff ourselves. Are there really not any utilities in chromium for this?

Reply:
for our specific use case of ecdh, i couldn't find anything useful from webcrypto

Reply:
or `crypt::Aead` with `AES_256_GCM?

Reply:
`crypto::ECPrivateKey`?

Reply:
i've tried using stuff from `components/webcrypto/algorithms` previously, but it ended up being a no-go since it required a few renderer specific blink dependencies

`crypt::Aead` should be an option, however. there may have been a reason why i chose not to use it but i can't recall. i'll revisit nevertheless.

the ecdh stuff would still remain, however

Reply:
ah, i remembered by `crypt::Aead` wasn't an option. we need the aes-128-gcm algo, and the class does not expose that algo. i suppose we could patch it to support it, but it probably wouldn't be worth it especially considering the simplicity of the aes portion of the code

Reply:
`ECPrivateKey` is also not useful since only EC points are exchanged between the server and client, and not PKCS#8/X.509 blocks which the `ECPrivateKey` uses for serialization.